File: harmonyos_samples/scan-kit_-sample-code_-clientdemo_-arkts/entry/src/main/ets/common/Logger.ts
/*
 * Copyright (c) 2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import hilog from '@ohos.hilog'

class Logger {
  private domain: number = 0x00001;
  private prefix: string = 'Scan SampleCode';

  debug(...args: string[]): void {
    let format: string = "%{public}s ".repeat(args.length);
    hilog.debug(this.domain, this.prefix, format, args);
  }

  info(...args: string[]): void {
    let format: string = "%{public}s ".repeat(args.length);
    hilog.info(this.domain, this.prefix, format, args);
  }

  warn(...args: string[]): void {
    let format: string = "%{public}s ".repeat(args.length);
    hilog.warn(this.domain, this.prefix, format, args);
  }

  error(...args: string[]): void {
    let format: string = "%{public}s ".repeat(args.length);
    hilog.error(this.domain, this.prefix, format, args);
  }
}

export default new Logger();

File: openharmony/arkui_ace_engine/frameworks/bridge/arkts_frontend/koala_projects/incremental/graphics/native/platform/cpp/android/logger.cc
/*
 * Copyright (c) 2022-2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <android/log.h>

#include <pthread.h>
#include <stdio.h>
#include <unistd.h>
#include "logdefs.h"

static int pfd[2];
static pthread_t thr;
static const char *tag = "koala";

static void *thread_func(void *) {
    ssize_t rdsz;
    char buf[128];
    while ((rdsz = read(pfd[0], buf, sizeof buf - 1)) > 0) {
        if (buf[rdsz - 1] == '\n') --rdsz;
        buf[rdsz] = 0;  /* add null-terminator */
        LOGE_ANDROID("%s", buf);
    }
    return nullptr;
}

int start_logger(const char *app_name) {
    tag = app_name;

    /* make stdout line-buffered and stderr unbuffered */
    setvbuf(stdout, nullptr, _IOLBF, 0);
    setvbuf(stderr, nullptr, _IONBF, 0);

    /* create the pipe and redirect stdout and stderr */
    pipe(pfd);
    dup2(pfd[1], 1);
    dup2(pfd[1], 2);

    /* spawn the logging thread */
    if (pthread_create(&thr, nullptr, thread_func, nullptr)) return -1;
    pthread_detach(thr);
    return 0;
}

